Adding certificate login (#13)
pnp/action-cli-login: README.md

## Usage
### Pre-requisites

Create a workflow `.yml` file in your `.github/workflows` directory. An [example workflow](#example-workflow---cli-for-microsoft-365-login) is available below. For more information, reference the GitHub Help Documentation for [Creating a workflow file](https://help.github.com/en/articles/configuring-a-workflow#creating-a-workflow-file).

### Inputs

> - `ADMIN_USERNAME` : Username (email address of the admin)

Reviewer: Lets change this to align with the action input description

```suggestion
- `ADMIN_USERNAME` : Username (upn) of the admin
```

- `ADMIN_PASSWORD` : Password of the admin
- `CERTIFICATE_ENCODED` : Base64-encoded string with certificate private key
- `CERTIFICATE_PASSWORD` : Password for the certificate

> - `AAD_APP_ID` : App ID of the Azure AD application to use for certificate authentication

Reviewer: Let's align this with the option name and description from `login` command for consistency.

```suggestion
- `APP_ID`: App ID of the Azure AD application to use for authentication. If not specified, use the app specified in the `CLIMICROSOFT365_AADAPPID` environment variable. If the environment variable is not defined, use the multitenant PnP Management Shell app
```

> - `TENANT_ID` : ID or domain (for example "contoso.onmicrosoft.com") of the tenant from which accounts should be able to authenticate

Reviewer: Let's align this with the option name and description from `login` command for consistency.

```suggestion
- `TENANT`: ID of the tenant from which accounts should be able to authenticate. Use `common` or `organization` if the app is multitenant. If not specified, use the tenant specified in the `CLIMICROSOFT365_TENANT` environment variable. If the environment variable is not defined, use `common` as the tenant identifier
```

All inputs are optional. But depending of the authentication type chosen, following pair of inputs will be necessary:

- `authType` is `password`: `ADMIN_USERNAME` and `ADMIN_PASSWORD` are required

> - `authType` is `certificate`: at least `CERTIFICATE_ENCODED` and `AAD_APP_ID` are required

Reviewer: ```suggestion
- `authType` is `certificate`: at least `CERTIFICATE_ENCODED` and `APP_ID` are required
```

>   - `TENANT_ID` will be required. It can be the ID of the tenant from which accounts should be able to authenticate. Use `common` or `organization` if the app is multitenant

Reviewer: ```suggestion
  - `TENANT` will be required. ID of the tenant from which accounts should be able to authenticate. Use `common` or `organization` if the app is multitenant
```

  - Depending on the certificate provided, if encoded with password, `CERTIFICATE_PASSWORD` will be required

#### Optional requirement

Since this action requires sensitive information such as user name, password and encoded certificate for example, it would be ideal to store them securely. We can achieve this in a GitHub repo by using [secrets](https://help.github.com/en/actions/automating-your-workflow-with-github-actions/creating-and-using-encrypted-secrets). So, click on `settings` tab in your repo and add:

- 2 new secrets if `authType` is `password`:
  - `ADMIN_USERNAME` - store the admin user name in this (e.g. user@contoso.onmicrosoft.com)
  - `ADMIN_PASSWORD` - store the password of that user in this.

> - 3 new secrets if `authType` is `certificate`:
>   - `AAD_APP_ID` - store the Azure AD application ID used for authentication
>   - `CERTIFICATE_ENCODED` - store the Base64-encoded string of the certificate stored in the Azure AD application
>   - `CERTIFICATE_PASSWORD` - store the certificate password

Reviewer: As `APP_ID` can also be used with password login, let's not suggest that it is just specific to certificate authentication.

Let's create a we have a section that calls out `APP_ID` and `TENANT` separately.

```suggestion
- 2 new secrets if `authType` is `certificate`:
  - `CERTIFICATE_ENCODED` - store the Base64-encoded string of the certificate stored in the Azure AD application
  - `CERTIFICATE_PASSWORD` - store the certificate password

- 2 new secrets if using a custom Azure AD identity:
  - `APP_ID` - store App ID of the Azure AD application to use for authentication
  - `TENANT` - store the ID of the tenant from which accounts should be able to authenticate
 
```

Create a workflow `.yml` file in your `.github/workflows` directory. An [example workflow](#example-workflow---cli-for-microsoft-365-login) is available below. For more information, reference the GitHub Help Documentation for [Creating a workflow file](https://help.github.com/en/articles/configuring-a-workflow#creating-a-workflow-file).

### Inputs
- `ADMIN_USERNAME` : Username (email address of the admin)

> - `PASSWORD` : Password for the admin or the certificate

Reviewer 2: If we rename `ADMIN_PASSWORD` to `PASSWORD`, wouldn't that be a breaking change?

Author: Sure yes. Would that be a problem?

Reviewer 2: Unless necessary, we should avoid breaking changes as it causes extra work for our users to use the latest version. In this case, I'd suggest that we use two separate variables even if it means a bit more work for us.

- `CERTIFICATE_ENCODED` : Base64-encoded string with certificate private key
- `AAD_APP_ID` : App ID of the Azure AD application to use for certificate authentication
- `TENANT_ID` : ID or domain (for example "contoso.onmicrosoft.com") of the tenant from which accounts should be able to authenticate

All inputs are optional. But depending of the authentication type chosen, following pair of inputs will be necessary:

- `authType` is `password`: `ADMIN_USERNAME` and `PASSWORD` are required
- `authType` is `certificate`: at least `CERTIFICATE_ENCODED` and `AAD_APP_ID` are required

>   - Depending on your tenant environment, `TENANT_ID` will be required

Reviewer 2: _Depending on your tenant environment_

Can we clarify this?

Author: Maybe I could copy / paste what is said on CLI for Microsoft 365 [login](https://pnp.github.io/cli-microsoft365/cmd/login/) documentation page, except the `CLIMICROSOFT365_TENANT` environment variable part? Like this:

```
ID of the tenant from which accounts should be able to authenticate. Use `common` or `organization` if the app is multitenant.
```

Reviewer 2: That would be better indeed 👍

  - Depending on the certificate provided, if encoded with password, `PASSWORD` will be required

#### Optional requirement
Since this action requires sensitive information such as user name, password and encoded certificate for example, it would be ideal to store them securely. We can achieve this in a GitHub repo by using [secrets](https://help.github.com/en/actions/automating-your-workflow-with-github-actions/creating-and-using-encrypted-secrets). So, click on `settings` tab in your repo and add:

- 2 new secrets if `authType` is `password`:
  - `adminUsername` - store the admin user name in this (e.g. user@contoso.onmicrosoft.com)
  - `adminPassword` - store the password of that user in this.

- 3 new secrets if `authType` is `certificate`:

>   - `AppID` - store the Azure AD application ID used for authentication

Reviewer 2: Is there any reason why we suggest naming first 2 secrets starting with lowercase and these 3 secrets with uppercase?

Author: No reason at all, just inconstitency 🤷‍♂️

Let me adjust this.

I didn't find any stated naming convention except the Actions REST API [code sample](https://docs.github.com/en/rest/reference/actions#list-organization-secrets--code-samples) which returns full uppercase secrets. So maybe we could follow this?

Reviewer 2: I'd say, let's align with the existing documentation so that we're not causing any breaking changes. We can always take a note of that to adjust in the next major version.

    # CLI for Microsoft 365 login action
    - name: Login to tenant

>       uses: pnp/action-cli-login@v3.0.0

Reviewer 2: The currently published action is v2.0.1. If we want to use a different version here, shouldn't we increase the version number in package.json?

Since we're adding an extra option to login without breaking the existing functionality, I'd suggest that we move to a v2.1.0 instead of v3.0.0.

Author: Oh I didn't see the version number on the package.json, I was only refering to the release notes in the README file.

The fact that all options are now optional (as they were required before) made me think about increasing major version. But I'm fine with increasing minor version.

      with:
        ADMIN_USERNAME:  ${{ secrets.adminUsername }}
        PASSWORD:  ${{ secrets.adminPassword }}
    
    ##
    ## Code to deploy the package to tenant omitted
    ##
```

### Example workflow - CLI for Microsoft 365 Login (certificate authentification)

> On every `push` build the code and then login to Office 365 before deploying, using certificate authentication.

Reviewer 2: Let's add an empty line between the heading an the paragraph

    ##
```

### Example workflow - CLI for Microsoft 365 Login (certificate authentification)

On every `push` build the code and then login to Office 365 before deploying, using certificate authentication.

```yaml
name: SPFx CICD with Cli for Microsoft 365

on: [push]

jobs:
  build:
    ##
    ## Build code omitted
    ##
        
  deploy:
    needs: build
    runs-on: ubuntu-latest
    strategy:
      matrix:
        node-version: [16.x]
    
    steps:
    
    ##
    ## Code to get the package omitted
    ##

    # CLI for Microsoft 365 login action
    - name: Login to tenant
      uses: pnp/action-cli-login@v2.1.0
      with:
        AAD_APP_ID:  ${{ secrets.appID }}
        CERTIFICATE_ENCODED: ${{ secrets.CERTIFICATE_ENCODED }}

>         CERTIFICATE_PASSWORD:  ${{ secrets.CERTIFICATE_PASSWORD }}

Reviewer: Let's also include the `tenant` option here in the the example as it is required for certificate authentication.

```suggestion
      with:
          TENANT: ${{secrets.TENANT }}
          APP_ID:  ${{ secrets.APP_ID }}
          CERTIFICATE_ENCODED: ${{ secrets.CERTIFICATE_ENCODED }}
          CERTIFICATE_PASSWORD:  ${{ secrets.CERTIFICATE_PASSWORD }}
```
